File: design/0432-all-oone-data-structure.go
package design

// https://leetcode.com/problems/all-oone-data-structure/

// Approach #2: Hash Map + Doubly Linked List
type allOneNode struct {
	key   string
	count int
	prev  *allOneNode
	next  *allOneNode
}

func NewAllOneNode(key string, prev *allOneNode, next *allOneNode) *allOneNode {
	return &allOneNode{key: key, prev: prev, next: next}
}

type AllOne struct {
	head *allOneNode
	tail *allOneNode
	keys map[string]*allOneNode
}

func NewAllOne() AllOne {
	head := NewAllOneNode("", nil, nil)
	head.count = 5e4
	tail := NewAllOneNode("", head, nil)
	head.next = tail

	return AllOne{
		keys: make(map[string]*allOneNode),
		head: head,
		tail: tail,
	}
}

func (a *AllOne) Inc(key string) {
	node := a.keys[key]
	if node == nil {
		node = NewAllOneNode(key, a.tail.prev, a.tail)
		node.prev.next = node
		node.next.prev = node
		a.keys[key] = node
	}

	node.count++
	for node.prev.count < node.count {
		a.swap(node.prev, node)
	}
}

func (a *AllOne) Dec(key string) {
	node := a.keys[key]
	node.count--

	if node.count == 0 {
		node.prev.next = node.next
		node.next.prev = node.prev
		a.keys[key] = nil

		return
	}

	for node.count < node.next.count {
		a.swap(node, node.next)
	}
}

func (a *AllOne) GetMaxKey() string {
	if a.head.next == a.tail {
		return ""
	}

	return a.head.next.key
}

func (a *AllOne) GetMinKey() string {
	if a.tail.prev == a.head {
		return ""
	}

	return a.tail.prev.key
}

func (a *AllOne) swap(node1 *allOneNode, node2 *allOneNode) {
	prev := node1.prev
	next := node2.next
	node1.prev, node2.prev = node2, prev
	node1.next, node2.next = next, node1
	prev.next = node2
	next.prev = node1
}

// // Approach #1: Trie
// type allOneNode struct {
// 	count    int
// 	children map[rune]*allOneNode
// }

// func NewAllOneNode() *allOneNode {
// 	return &allOneNode{children: make(map[rune]*allOneNode)}
// }

// type AllOne struct {
// 	root *allOneNode
// }

// func NewAllOne() AllOne {
// 	return AllOne{root: NewAllOneNode()}
// }

// func (a *AllOne) Inc(key string) {
// 	curr := a.root
// 	curr.count++
// 	for _, c := range key {
// 		if curr.children[c] == nil {
// 			curr.children[c] = NewAllOneNode()
// 		}
// 		curr = curr.children[c]
// 		curr.count++
// 	}
// }

// func (a *AllOne) Dec(key string) {
// 	curr := a.root
// 	curr.count--
// 	for _, c := range key {
// 		curr = curr.children[c]
// 		curr.count--
// 	}
// }

// func (a *AllOne) GetMaxKey() string {
// 	var key = ""
// 	curr := a.root
// 	var next *allOneNode
// 	var nextc string
// 	var count int
// 	for curr != nil {
// 		next = nil
// 		nextc = ""
// 		count = 0
// 		for c, child := range curr.children {
// 			if child.count > count {
// 				count = child.count
// 				next = child
// 				nextc = string(c)
// 			}
// 		}
// 		curr = next
// 		key += nextc
// 	}
// 	return key
// }

// func (a *AllOne) GetMinKey() string {
// 	var key = ""
// 	var curr = a.root
// 	var next *allOneNode
// 	var nextc string
// 	var count int
// 	for curr != nil {
// 		next = nil
// 		nextc = ""
// 		count = 5e4
// 		for c, child := range curr.children {
// 			if child.count == 0 {
// 				continue
// 			}
// 			if child.count < count {
// 				count = child.count
// 				next = child
// 				nextc = string(c)
// 			}
// 		}
// 		curr = next
// 		key += nextc
// 	}
// 	return key
// }

/**
 * Your AllOne object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Inc(key);
 * obj.Dec(key);
 * param_3 := obj.GetMaxKey();
 * param_4 := obj.GetMinKey();
 */

File: design/0432-all-oone-data-structure.py
from typing import List
import unittest

# https://leetcode.com/problems/all-oone-data-structure/

# python3 -m unittest design/0432-all-oone-data-structure.py

# # Approach: Trie
# class TrieNode:
#     def __init__(self, val: str):
#         self.val = val
#         self.count = 0
#         self.children = {}

# class AllOne:
#     def __init__(self):
#         self.root = TrieNode(0)

#     # Time: O(len(key))
#     def inc(self, key: str) -> None:
#         curr = self.root
#         curr.count += 1
#         for c in key:
#             if c not in curr.children:
#                 curr.children[c] = TrieNode(c)
#             curr = curr.children[c]
#             curr.count += 1

#     # Time: O(len(key))
#     def dec(self, key: str) -> None:
#         curr = self.root
#         curr.count -= 1
#         for c in key:
#             curr = curr.children[c]
#             curr.count -= 1

#     # Time: O(26*len(key))
#     def getKey(self, first_count, compare) -> str:
#         key = ""
#         curr = self.root
#         while curr:
#             nxt = None
#             ch = ''
#             count = first_count
#             for c, child in curr.children.items():
#                 if child.count == 0:
#                     continue
#                 if compare(child.count, count):
#                     count = child.count
#                     nxt = child
#                     ch = c
#             key += ch
#             curr = nxt
#         return key

#     def getMaxKey(self) -> str:
#         return self.getKey(0, lambda a, b: a > b)

#     def getMinKey(self) -> str:
#         return self.getKey(int(5e4), lambda a, b: a < b)


# Approach: Hash Map + Doubly Linked List
class ListNode:
    def __init__(self, key, prev: "ListNode", nxt: "ListNode"):
        self.key = key
        self.count = 0
        self.prev = prev
        self.nxt = nxt


class AllOne:
    def __init__(self):
        self.keys = {}
        self.head = ListNode("", None, None)
        self.head.count = 5e4
        self.tail = ListNode("", self.head, None)
        self.head.nxt = self.tail

    def swap(self, nodes1: List[ListNode], nodes2: List[ListNode]) -> None:
        node1, node2 = nodes1[0], nodes2[0]
        prev = node1.prev
        nxt = node2.nxt
        node1.prev, node2.prev = node2, prev
        node1.nxt, node2.nxt = nxt, node1
        prev.nxt = node2
        nxt.prev = node1

    def inc(self, key: str) -> None:
        if key not in self.keys:
            node = ListNode(key, self.tail.prev, self.tail)
            node.prev.nxt = node
            node.nxt.prev = node
            self.keys[key] = node
        node = self.keys[key]
        node.count += 1
        while node.prev.count < node.count:
            self.swap([node.prev], [node])

    def dec(self, key: str) -> None:
        node = self.keys[key]
        node.count -= 1
        if node.count == 0:
            node.prev.nxt = node.nxt
            node.nxt.prev = node.prev
            del self.keys[key]
            return
        while node.count < node.nxt.count:
            self.swap([node], [node.nxt])

    def getMaxKey(self) -> str:
        if self.head.nxt == self.tail:
            return ""
        return self.head.nxt.key

    def getMinKey(self) -> str:
        if self.tail.prev == self.head:
            return ""
        return self.tail.prev.key


class TestAllOne(unittest.TestCase):
    def test(self):
        for commands, values, expected in [
            (
                ["AllOne", "inc", "inc", "getMaxKey", "getMinKey", "inc", "getMaxKey", "getMinKey"],
                [[], ["hello"], ["hello"], [], [], ["leet"], [], []],
                [None, None, None, "hello", "hello", None, "hello", "leet"],
            ),
            (
                [
                    "AllOne",
                    "inc",
                    "inc",
                    "inc",
                    "inc",
                    "getMaxKey",
                    "inc",
                    "inc",
                    "inc",
                    "dec",
                    "inc",
                    "inc",
                    "inc",
                    "getMaxKey",
                ],
                [
                    [],
                    ["hello"],
                    ["goodbye"],
                    ["hello"],
                    ["hello"],
                    [],
                    ["leet"],
                    ["code"],
                    ["leet"],
                    ["hello"],
                    ["leet"],
                    ["code"],
                    ["code"],
                    [],
                ],
                [None, None, None, None, None, "hello", None, None, None, None, None, None, None, "leet"],
            ),
            (
                [
                    "AllOne",
                    "inc",
                    "inc",
                    "inc",
                    "inc",
                    "inc",
                    "inc",
                    "dec",
                    "dec",
                    "getMinKey",
                    "dec",
                    "getMaxKey",
                    "getMinKey",
                ],
                [[], ["a"], ["b"], ["b"], ["c"], ["c"], ["c"], ["b"], ["b"], [], ["a"], [], []],
                [None, None, None, None, None, None, None, None, None, "a", None, "c", "c"],
            ),
        ]:
            for idx, command in enumerate(commands):
                output = None
                match command:
                    case "AllOne":
                        obj = AllOne()
                    case "inc":
                        obj.inc(values[idx][0])
                    case "dec":
                        obj.dec(values[idx][0])
                    case "getMaxKey":
                        output = obj.getMaxKey()
                    case "getMinKey":
                        output = obj.getMinKey()
                self.assertEqual(expected[idx], output, f"expected: {expected[idx]}, output: {output}")


# Your AllOne object will be instantiated and called as such:
# obj = AllOne()
# obj.inc(key)
# obj.dec(key)
# param_3 = obj.getMaxKey()
# param_4 = obj.getMinKey()
